File: language/transaction-builder/generator/examples/java/StdlibDemo.java
// Copyright (c) The Diem Core Contributors
// SPDX-License-Identifier: Apache-2.0

import java.util.Arrays;
import java.util.ArrayList;

import com.novi.serde.Bytes;
import com.novi.serde.Unsigned; // used as documentation.
import com.diem.stdlib.Helpers;
import com.diem.stdlib.ScriptCall;;
import com.diem.stdlib.ScriptFunctionCall;;
import com.diem.types.AccountAddress;
import com.diem.types.Identifier;
import com.diem.types.Script;
import com.diem.types.StructTag;
import com.diem.types.TypeTag;
import com.diem.types.TransactionPayload;

public class StdlibDemo {
    private static void demo_p2p_script() throws Exception {
        StructTag.Builder builder = new StructTag.Builder();
        builder.address = AccountAddress.valueOf(new byte[]{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1});
        builder.module = new Identifier("XDX");
        builder.name = new Identifier("XDX");
        builder.type_params = new ArrayList<com.diem.types.TypeTag>();
        StructTag tag = builder.build();

        TypeTag token = new TypeTag.Struct(tag);

        AccountAddress payee = AccountAddress.valueOf(
            new byte[]{0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22});

        @Unsigned Long amount = Long.valueOf(1234567);
        Script script =
            Helpers.encode_peer_to_peer_with_metadata_script(token, payee, amount, Bytes.empty(), Bytes.empty());

        ScriptCall.PeerToPeerWithMetadata call = (ScriptCall.PeerToPeerWithMetadata)Helpers.decode_script(script);
        assert(call.amount.equals(amount));
        assert(call.payee.equals(payee));

        byte[] output = script.bcsSerialize();
        for (byte o : output) {
            System.out.print(((int) o & 0xFF) + " ");
        };
        System.out.println();
    }

    private static void demo_p2p_script_function() throws Exception {
        StructTag.Builder builder = new StructTag.Builder();
        builder.address = AccountAddress.valueOf(new byte[]{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1});
        builder.module = new Identifier("XDX");
        builder.name = new Identifier("XDX");
        builder.type_params = new ArrayList<com.diem.types.TypeTag>();
        StructTag tag = builder.build();

        TypeTag token = new TypeTag.Struct(tag);

        AccountAddress payee = AccountAddress.valueOf(
            new byte[]{0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22});

        @Unsigned Long amount = Long.valueOf(1234567);
        TransactionPayload payload =
            Helpers.encode_peer_to_peer_with_metadata_script_function(token, payee, amount, Bytes.empty(), Bytes.empty());

        ScriptFunctionCall.PeerToPeerWithMetadata call = (ScriptFunctionCall.PeerToPeerWithMetadata)Helpers.decode_script_function_payload(payload);
        assert(call.amount.equals(amount));
        assert(call.payee.equals(payee));

        byte[] output = payload.bcsSerialize();
        for (byte o : output) {
            System.out.print(((int) o & 0xFF) + " ");
        };
        System.out.println();
    }

    public static void main(String[] args) throws Exception {
        demo_p2p_script();
        demo_p2p_script_function();
    }
}

File: language/transaction-builder/generator/examples/csharp/StdlibDemo.cs
using System;
using System.Collections.Generic;
using Diem.Stdlib;
using Diem.Types;
using Serde;
using System.Diagnostics;

public class StdlibDemo {
    public static void Main() {
        DemoP2PScript();
        DemoP2PScriptFunction();
    }

    static void DemoP2PScript() {
        StructTag tag = new StructTag(
            AccountAddress.valueOf(new byte[]{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1}),
            new Identifier("XDX"),
            new Identifier("XDX"),
            new ValueArray<TypeTag>(new List<TypeTag>().ToArray())
        );

        TypeTag token = new TypeTag.Struct(tag);

        AccountAddress payee = AccountAddress.valueOf(
            new byte[]{0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22});

        ulong amount = 1234567;
        Script script =
            Helpers.encode_peer_to_peer_with_metadata_script(
                token,
                payee,
                amount,
                new ValueArray<byte>(new List<byte>().ToArray()),
                new ValueArray<byte>(new List<byte>().ToArray())
        );

        ScriptCall.PeerToPeerWithMetadata call = (ScriptCall.PeerToPeerWithMetadata)Helpers.DecodeScript(script);
        Debug.Assert(call.amount.Equals(amount), string.Format("call.amount is {0}. Expecting {1}", call.amount, amount));
        Debug.Assert(call.payee.Equals(payee), string.Format("call.payee is {0}. Expecting {1}", call.payee, payee));

        byte[] output = script.BcsSerialize();
        foreach (byte o in output) {
            Console.Write(((int) o & 0xFF) + " ");
        }
        Console.WriteLine();
    }

    static void DemoP2PScriptFunction() {
        StructTag tag = new StructTag(
            AccountAddress.valueOf(new byte[]{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1}),
            new Identifier("XDX"),
            new Identifier("XDX"),
            new ValueArray<TypeTag>(new List<TypeTag>().ToArray())
        );

        TypeTag token = new TypeTag.Struct(tag);

        AccountAddress payee = AccountAddress.valueOf(
            new byte[]{0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22});

        ulong amount = 1234567;
        TransactionPayload payload =
            Helpers.encode_peer_to_peer_with_metadata_script_function(
                token,
                payee,
                amount,
                new ValueArray<byte>(new List<byte>().ToArray()),
                new ValueArray<byte>(new List<byte>().ToArray())
        );

        ScriptFunctionCall.PeerToPeerWithMetadata call = (ScriptFunctionCall.PeerToPeerWithMetadata)Helpers.DecodeScriptFunctionPayload(payload);
        Debug.Assert(call.amount.Equals(amount), string.Format("call.amount is {0}. Expecting {1}", call.amount, amount));
        Debug.Assert(call.payee.Equals(payee), string.Format("call.payee is {0}. Expecting {1}", call.payee, payee));

        byte[] output = payload.BcsSerialize();
        foreach (byte o in output) {
            Console.Write(((int) o & 0xFF) + " ");
        }
        Console.WriteLine();
    }
}
